### PR TITLE
3. Send notfound when Zebra doesn't have a block or transaction

### LINKED ISSUE #2726: Make Zebra send `notfound`, and use received `notfound` to finish requests
## Motivation

`zcashd` and Zebra don't actually need `notfound` messages:
https://github.com/zcash/zcash/blob/e7b425298f6d9a54810cb7183f00be547e4d9415/src/main.cpp#L5632
https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer/connection.rs#L115

But Zebra should send `notfound` as much as possible, so that `inv` routing works efficiently.
This also helps with testing #2156.

We might also want to use them for SPV clients - does the Zcash network have any of those?

### Suggested Solution

- send `notfound` so Zebra peers know that we've replied with all the transactions/blocks we have
- make sure Zebra checks for `notfound` when it's waiting for more blocks or transactions (we might already do this)

### Rejected Alternatives

`zcashd` doesn't do this, and it isn't really supported by the protocol.

<details>

- send `notfound` to tell peers when our mempool is empty
  - make Zebra check for `notfound` when waiting for a `MempoolTransactionIds` reply, rather than timing out because there is no reply

</details>

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Currently, Zebra just ignores block or transaction requests it can't answer.
But we should tell peers we don't have that inventory, so they can try another peer.

## Solution

- send notfound when Zebra doesn't have a block or transaction
- allow different available and missing types inside ResponseStatus
- stop sending unsupported missing inventory types to the registry

Related non-feature fixes:
- move address sanitization into the response future
- add a block_hash convenience method to InventoryHash
- improve failure logs for connection tests
- move module docs to the top of each module

Tests:
- [x] inbound messages are forwarded to the registry
- [x] Zebra responds with `notfound` block messages to peers - single block request
- [x] Zebra responds with `notfound` transaction messages to peers (including Tx and Wtx) - multi-transaction request

Closes #2726.

## Review

I'd like reviews to focus on the functionality changes.
(I will open another PR to deal with some edge cases around peers not sending `notfound`, and Zebra cancelling requests.)

@jvff can review this PR.

This PR is based on PR #3465.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

- #2156 
    - [ ] if a block or transaction request fails, times out, or is dropped, send a synthetic `notfound` message to the inv collector

### CODE REVIEW

File: zebrad/src/components/inbound.rs [hunk 1]
```diff
@@ -360,22 +375,36 @@ impl Service<zn::Request> for Inbound {
...
-                            zn::Response::Nil
-                        } else {
-                            zn::Response::Blocks(blocks)
-                        }
+                        // Work out which hashes were missing.
+                        let available_hashes: HashSet<block::Hash> = blocks.iter().map(|block| block.hash()).collect();
+                        let available = blocks.into_iter().map(Available);
+                        let missing = hashes.into_iter().filter(|hash| !available_hashes.contains(hash)).map(Missing);
+
+                        zn::Response::Blocks(available.chain(missing).collect())
```
Comment (teor2345) [inline zebrad/src/components/inbound.rs:383]: I still need to write tests for this, but I think they should be pretty simple.

Do we have any tests that run against a full Zebra inbound service, network stack, and empty state?

(We can mock the state, but the inbound service and zebra-network need to be real.)

Comment (teor2345) [reply zebrad/src/components/inbound.rs:383]: This seems like a good place to start:
https://github.com/ZcashFoundation/zebra/blob/1a14baf8e7c6c82dba7a64e9b43d076b8f856f2a/zebrad/src/components/inbound/tests.rs#L655

**Comment by teor2345** [comment]
@jvff this should be ready for review now.

I'd like to focus on the functionality changes, because this PR series has taken longer than I expected.
So if we have to leave cleanups and refactors for later, that's ok.

**Comment by teor2345** [comment]
Failed due to #3489:
https://github.com/ZcashFoundation/zebra/runs/5115656563?check_suite_focus=true#step:10:1148

File: zebra-network/src/peer/minimum_peer_version.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+//! Watches for chain tip height updates to determine the minimum support peer protocol version.
```
Comment (jvff) [inline zebra-network/src/peer/minimum_peer_version.rs]: Optional: I just realized something that might be a typo: should this be `supported`?

```suggestion
//! Watches for chain tip height updates to determine the minimum supported peer protocol version.
```

Comment (teor2345) [reply zebra-network/src/peer/minimum_peer_version.rs]: I moved an existing module doc to avoid duplicates and inconsistencies.

It seems like a typo that's easy to fix.

File: zebra-network/src/protocol/internal/response_status.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+use ResponseStatus::*;
+
+/// A generic peer inventory response status.
+///
+/// `Available` is used for inventory that is present in the response,
+/// and `Missing` is used for inventory that is missing from the response.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub enum ResponseStatus<A, M> {
```
Comment (jvff) [inline zebra-network/src/protocol/internal/response_status.rs:16]: Nit: I'm a bit worried if `ResponseStatus` might be a confusing name. My first thought was that it could be misinterpreted as "status of a response". I'd suggest something like:

- `InventoryResponse`
- `InventoryStatusResponse`
- `InvResponse`
- `StatusResponse`
- ...?

We can also open an issue for this and do this later if you'd prefer.

File: zebra-network/src/protocol/internal/response_status.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    /// Get the missing inventory item, if present.
+    #[allow(dead_code)]
+    pub fn missing(&self) -> Option<M> {
+        if let Missing(item) = self {
+            Some(item.clone())
+        } else {
+            None
+        }
+    }
+}
```
Comment (jvff) [inline zebra-network/src/protocol/internal/response_status.rs:101]: Optional: usually I'd expect these methods to return references instead of clones, but I'm not sure if it would impact ergonomics too much in other code that use them :thinking:

File: zebra-network/src/peer_set/inventory_registry.rs [hunk 5]
```diff
@@ -125,64 +118,27 @@ impl InventoryChange {
...
             Missing(item) => Missing(f(item)),
         }
     }
+}
 
-    /// Converts from `&InventoryStatus<T>` to `InventoryStatus<&T>`.
-    pub fn as_ref(&self) -> InventoryStatus<&T> {
+impl<T: Clone> InventoryStatus<T> {
+    /// Get the inner item, regardless of status.
+    pub fn inner(&self) -> T {
```
Comment (jvff) [inline zebra-network/src/peer_set/inventory_registry.rs:139]: Optional: I'd suggest renaming this `to_inner`, so that it gives a hint that the type is being cloned (inspired by the [Rust API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)).

```suggestion
    pub fn to_inner(&self) -> T {
```

File: zebra-network/src/peer_set/inventory_registry/tests/prop.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    #[test]
+    fn inv_registry_inbound_wrapper_ok(
+        status in any::<InventoryMarker>(),
+        test_hashes in prop::collection::hash_set(any::<InventoryHash>(), 0..=MAX_HASH_COUNT)
+    ) {
+        prop_assert!(MAX_HASH_COUNT <= MAX_PENDING_CHANGES, "channel must not block in tests");
+
+        // Start the runtime
+        let runtime = zebra_test::init_async();
+        let _guard = runtime.enter();
```
Comment (jvff) [inline zebra-network/src/peer_set/inventory_registry/tests/prop.rs:32]: Minor: The `runtime.enter()` is only required if the code that's outside the future requires using Tokio (like for example spawning tasks or setting timers). In this case it could be removed:

```suggestion
```

File: zebrad/src/components/inbound.rs [hunk 7]
```diff
@@ -307,33 +313,41 @@ impl Service<zn::Request> for Inbound {
...
-
-                if !peers.is_empty() {
-                    async { Ok(zn::Response::Peers(peers)) }.boxed()
-                } else {
-                    debug!("ignoring `Peers` request from remote peer because our address book is empty");
-                    async { Ok(zn::Response::Nil) }.boxed()
-                }
+                async move {
+                    // Correctness: get the current time after acquiring the address book lock.
+                    let now = Utc::now();
```
Comment (jvff) [inline zebrad/src/components/inbound.rs:322]: Question: should `now` indicate the time that address book snapshot was taken, or the time the future to send the response was executed?

File: zebrad/src/components/inbound.rs [hunk 8]
```diff
@@ -344,6 +358,7 @@ impl Service<zn::Request> for Inbound {
                 // https://github.com/tower-rs/tower/blob/master/tower/src/util/call_all/common.rs#L112
                 use futures::stream::TryStreamExt;
                 hashes
+                    .clone()
```
Comment (jvff) [inline zebrad/src/components/inbound.rs:361]: Minor: I'd suggest replacing `hashes.clone().into_iter()` with `hashes.iter().cloned()`, so that we avoid allocating space for a temporary container and maybe also avoid rehashing the items.

File: zebra-network/src/peer/connection.rs [hunk 9]
```diff
@@ -286,15 +287,21 @@ impl Handler {
                     //
                     // Ignoring the message gives us a chance to synchronize back to the correct
                     // request.
+                    //
+                    // Peers can avoid these cascading errors by sending an explicit `notfound`.
+                    // Zebra sends `notfound`, but `zcashd` doesn't.
```
Comment (jvff) [inline zebra-network/src/peer/connection.rs:292]: I'm not sure this is related to the cancel request error cascade. If the local request is cancelled, there's no way to notify the remote peer so that it can send us a `notfound` message.

File: zebra-network/src/peer/connection.rs [hunk 10]
```diff
@@ -1216,18 +1218,48 @@ where
...
+                    if let Err(e) = self.peer_tx.send(Message::NotFound(missing_ids)).await {
                         self.fail_with(e);
                         return;
                     }
                 }
             }
             Response::Blocks(blocks) => {
-                // Generate one block message per block.
+                // Generate one tx message per block,
+                // then a notfound% message with all the missing block hashes.
```
Comment (jvff) [inline zebra-network/src/peer/connection.rs:1246]: Optional: typo?

```suggestion
                // then a notfound message with all the missing block hashes.
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+        tx_gossip_result,
+    );
+
+    Ok(())
+}
+
+/// Setup a real Zebra network stack, with a connected peer using a real isolated network stack.
+///
+/// Uses fake verifiers, and does not run a block syncer task.
+async fn setup() -> (
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:326]: The amount of items prepared during test setup suggest to me that a some sort of test harness type would help here. But that would likely take some time to think through and get it working.

Is it okay for me to open an issue with priority:optional for this so we can revisit it in the future?

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+        // real tasks
+        block_gossip_task_handle,
+        tx_gossip_task_handle,
+        // real open socket addresses
+        listen_addr,
+    ) = setup().await;
+
+    // Use inbound directly
+    let request = inbound_service.clone().oneshot(Request::Peers);
+    let response = request.await;
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:64]: Nit: Having two bindings that have paired names but non-paired types seem confusing to me (i.e., one is a `Future` doing a request, the other is an actual `Response` type).

```suggestion
    let response = inbound_service.clone().oneshot(Request::Peers).await;
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+        _ => unreachable!(
+            "`Peers` requests should always respond `Ok(Response::Peers(_))`, \
+             actual result: {:?}",
+            response
+        ),
+    };
+
+    // Use the connected peer via a local TCP connection
+    let request = connected_peer_service.clone().oneshot(Request::Peers);
+    let response = request.await;
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:83]: Nit: Same as above.

```suggestion
    let response = connected_peer_service.clone().oneshot(Request::Peers).await;
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+        _listen_addr,
+    ) = setup().await;
+
+    let test_block = block::Hash([0x11; 32]);
+
+    // Use inbound directly
+    let request = inbound_service
+        .clone()
+        .oneshot(Request::BlocksByHash(iter::once(test_block).collect()));
+    let response = request.await;
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:145]: Nit: Same as above.

```suggestion
    let response = inbound_service
        .clone()
        .oneshot(Request::BlocksByHash(iter::once(test_block).collect()))
        .await;
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+             actual result: {:?}",
+            response
+        ),
+    };
+
+    // Use the connected peer via a local TCP connection
+    let request = connected_peer_service
+        .clone()
+        .oneshot(Request::BlocksByHash(iter::once(test_block).collect()));
+    let response = request.await;
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:166]: Nit: Same as above:

```suggestion
    let response = connected_peer_service
        .clone()
        .oneshot(Request::BlocksByHash(iter::once(test_block).collect()))
        .await;
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+    }
+    .into();
+
+    // Test both transaction ID variants, separately and together
+    for txs in [vec![test_tx], vec![test_wtx], vec![test_tx, test_wtx]] {
+        // Use inbound directly
+        let request = inbound_service
+            .clone()
+            .oneshot(Request::TransactionsById(txs.iter().copied().collect()));
+        let response = request.await;
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:238]: Nit: Same as above.

```suggestion
        let response = inbound_service
            .clone()
            .oneshot(Request::TransactionsById(txs.iter().copied().collect()))
            .await;
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+                 actual result: {:?}",
+                response
+            ),
+        };
+
+        // Use the connected peer via a local TCP connection
+        let request = connected_peer_service
+            .clone()
+            .oneshot(Request::TransactionsById(txs.iter().copied().collect()));
+        let response = request.await;
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:263]: Nit: Same as above.

```suggestion
        let response = connected_peer_service
            .clone()
            .oneshot(Request::TransactionsById(txs.iter().copied().collect()))
            .await;
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+    );
+
+    Ok(())
+}
+
+/// Check that a network stack with an empty state responds to single transaction requests with `notfound`.
+///
+/// Uses a real Zebra network stack, with an isolated Zebra inbound TCP connection.
+///
+/// TODO: test a response with some Available and some Missing transactions.
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:205]: Question: is there an issue open to track this?

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+
+use ResponseStatus::*;
+
+/// Check that a network stack with an empty address book only contains the local listener port,
+/// by querying the inbound service via a local TCP connection.
+///
+/// Uses a real Zebra network stack with a local listener address,
+/// and an isolated Zebra inbound TCP connection.
+#[tokio::test]
+async fn inbound_peers_empty_address_book() -> Result<(), crate::BoxError> {
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:44]: Minor: I think the function doesn't need to return a `Result`:

```suggestion
async fn inbound_peers_empty_address_book() {
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+    );
+
+    let tx_gossip_result = tx_gossip_task_handle.now_or_never();
+    assert!(
+        matches!(tx_gossip_result, None),
+        "unexpected error or panic in transaction gossip task: {:?}",
+        tx_gossip_result,
+    );
+
+    Ok(())
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:114]: Minor: ...continuing removal of `Result` return.

```suggestion
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 21]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+/// Check that a network stack with an empty state responds to block requests with `notfound`.
+///
+/// Uses a real Zebra network stack, with an isolated Zebra inbound TCP connection.
+#[tokio::test]
+async fn inbound_block_empty_state_notfound() -> Result<(), crate::BoxError> {
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:121]: Minor: Same as above.

```suggestion
async fn inbound_block_empty_state_notfound() {
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 22]
```diff
@@ -0,0 +1,498 @@
...
+    );
+
+    let tx_gossip_result = tx_gossip_task_handle.now_or_never();
+    assert!(
+        matches!(tx_gossip_result, None),
+        "unexpected error or panic in transaction gossip task: {:?}",
+        tx_gossip_result,
+    );
+
+    Ok(())
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:198]: Minor: ...continued

```suggestion
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 23]
```diff
@@ -0,0 +1,498 @@
...
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:320]: Minor: ...continued.

```suggestion
```

File: zebrad/src/components/inbound/tests/real_peer_set.rs [hunk 24]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+/// Check that a network stack with an empty state responds to single transaction requests with `notfound`.
+///
+/// Uses a real Zebra network stack, with an isolated Zebra inbound TCP connection.
+///
+/// TODO: test a response with some Available and some Missing transactions.
+#[tokio::test]
+async fn inbound_tx_empty_state_notfound() -> Result<(), crate::BoxError> {
```
Comment (jvff) [inline zebrad/src/components/inbound/tests/real_peer_set.rs:207]: Nit: same as above.

```suggestion
async fn inbound_tx_empty_state_notfound() {
```

**Review by jvff — DISMISSED** [review]
I left a lot of comments, but I don't think any of them should block merging the PR 🤔

File: zebra-network/src/protocol/internal/response_status.rs (hunk 4 above)
Comment (teor2345) [reply zebra-network/src/protocol/internal/response_status.rs:101]: These methods are used on `Copy` types, so any clones are probably optimised out.

If they ever appear in benchmarks or flamegraphs, happy to fix them then.

File: zebra-network/src/peer/connection.rs (hunk 9 above)
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:292]: Good question!

It's even worse than that, `zcashd` doesn't send `notfound` in response to block requests. (It only sends `notfound` for transactions.)

This is fixed by the `MissingInventoryCollector` in the next PR, specifically:
https://github.com/ZcashFoundation/zebra/pull/3505/files#diff-946777883d52596b10e4679ed99713f02c8afe589c8ac641015b473d82b6dd56R366-R409

File: zebrad/src/components/inbound/tests/real_peer_set.rs (hunk 11 above)
Comment (teor2345) [reply zebrad/src/components/inbound/tests/real_peer_set.rs:326]: I'm not sure.

I asked "which tasks are worth creating?" on the draft Process document. I think we'll have a clearer answer after that discussion.

File: zebrad/src/components/inbound/tests/real_peer_set.rs (hunk 18 above)
Comment (teor2345) [reply zebrad/src/components/inbound/tests/real_peer_set.rs:205]: This case will be covered in PR #3505.

File: zebrad/src/components/inbound/tests/real_peer_set.rs (hunk 12 above)
Comment (teor2345) [reply zebrad/src/components/inbound/tests/real_peer_set.rs:64]: I don't think these bindings will cause much confusion, particularly in the tests.

Next time you want to make a large number of similar suggestions, feel free to open a fix PR. (If you decide to do it for these changes, please base it on PR #3505.)

File: zebrad/src/components/inbound/tests/real_peer_set.rs (hunk 19 above)
Comment (teor2345) [reply zebrad/src/components/inbound/tests/real_peer_set.rs:44]: I don't think this matters that much in tests.

File: zebra-network/src/protocol/internal/response_status.rs (hunk 3 above)
Comment (teor2345) [reply zebra-network/src/protocol/internal/response_status.rs:16]: Fixed in 30212dadf in PR #3505

File: zebra-network/src/peer_set/inventory_registry.rs (hunk 5 above)
Comment (teor2345) [reply zebra-network/src/peer_set/inventory_registry.rs:139]: Fixed in 7c32990c3 in PR #3505

File: zebra-network/src/peer_set/inventory_registry/tests/prop.rs (hunk 6 above)
Comment (teor2345) [reply zebra-network/src/peer_set/inventory_registry/tests/prop.rs:32]: Fixed in d10d52859 in PR #3505

File: zebrad/src/components/inbound.rs (hunk 7 above)
Comment (teor2345) [reply zebrad/src/components/inbound.rs:322]: It doesn't matter, it's just used to filter outdated peers.

Documented in c7914a48f in PR #3505

File: zebrad/src/components/inbound.rs (hunk 8 above)
Comment (teor2345) [reply zebrad/src/components/inbound.rs:361]: We can get hundreds of thousands of these requests if another node is syncing from us. But most requests from `zebrad` only contain one block. (I'm not sure what `zcashd` does.)

So I don't know if this optimisation actually matters.

Fixed in e3914c957 in PR #3505

File: zebra-network/src/peer/connection.rs (hunk 10 above)
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:1246]: Fixed in 88fe48693 in PR #3505

File: zebra-network/src/peer/connection.rs (hunk 9 above)
Comment (jvff) [reply zebra-network/src/peer/connection.rs:292]: Oh, I see. I'd suggest rewording this to be clearer:

```suggestion
                    // Peers can avoid these cascading errors in scenario #2 by sending an explicit
                    // `notfound`. Zebra sends `notfound`, but `zcashd` doesn't.
```

File: zebra-network/src/protocol/internal/response_status.rs [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+    /// Returns the response status type as a string.
+    pub fn command(&self) -> &'static str {
+        match self {
+            ResponseStatus::Available(_) => "Available",
+            ResponseStatus::Missing(_) => "Missing",
+        }
+    }
+
+    /// Returns true if the inventory item was available.
+    #[allow(dead_code)]
```
Comment (jvff) [inline zebra-network/src/protocol/internal/response_status.rs:40]: Minor: This could be removed.

```suggestion
```

File: zebra-network/src/protocol/internal/response_status.rs [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+    pub fn is_available(&self) -> bool {
+        matches!(self, Available(_))
+    }
+
+    /// Returns true if the inventory item was missing.
+    #[allow(dead_code)]
```
Comment (jvff) [inline zebra-network/src/protocol/internal/response_status.rs:46]: Minor: This could be removed.

```suggestion
```

File: zebra-network/src/protocol/internal/response_status.rs [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+    pub fn is_missing(&self) -> bool {
+        matches!(self, Missing(_))
+    }
+
+    /// Maps a `ResponseStatus<A, M>` to `ResponseStatus<B, M>` by applying a function to a
+    /// contained [`Available`] value, leaving the [`Missing`] value untouched.
+    #[allow(dead_code)]
```
Comment (jvff) [inline zebra-network/src/protocol/internal/response_status.rs:53]: Minor: This could be removed.

```suggestion
```

File: zebra-network/src/protocol/internal/response_status.rs [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+        // Based on Result::map from https://doc.rust-lang.org/src/core/result.rs.html#765
+        match self {
+            Available(a) => Available(f(a)),
+            Missing(m) => Missing(m),
+        }
+    }
+
+    /// Maps a `ResponseStatus<A, M>` to `ResponseStatus<A, N>` by applying a function to a
+    /// contained [`Missing`] value, leaving the [`Available`] value untouched.
+    #[allow(dead_code)]
```
Comment (jvff) [inline zebra-network/src/protocol/internal/response_status.rs:64]: Minor: This could be removed.

```suggestion
```

File: zebra-network/src/protocol/internal/response_status.rs (hunk 25 above)
Comment (teor2345) [reply zebra-network/src/protocol/internal/response_status.rs:40]: Fixed in 0c5f18ddf, along with all the other similar issues

**Comment by teor2345** [comment]
@jvff thanks for the reviews.

This PR series took much longer than I expected to write and test, so I'd really like to get it merged.
Are there any major issues with the design, or any bugs you can see?

**Review by jvff — APPROVED** [review]

File: zebra-network/src/peer/connection.rs (hunk 9 above)
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:292]: Expanded in 961edc051, I think it's worth writing a bit more to clarify this scenario, because we know `zcashd` doesn't send `notfound`.
